### PR TITLE
feat(weave): ImperativeEval in TS SDK Part 1

### PR DESCRIPTION
**Pull request by chance-wnb**
# Imperative Evaluation Logger for TypeScript SDK

## JIRA:

https://wandb.atlassian.net/browse/WB-28160

## Summary
This PR implements **Imperative Evaluation Logger** for the TypeScript SDK, providing an alternative to the traditional batch-oriented `Evaluation` class. This feature allows incremental logging of predictions and scores as they happen, without requiring a fixed dataset or batch processing.

This implementation achieves feature parity with the Python SDK's `eval_imperative.py`.

## Motivation

**This feature already exists in the Python SDK but was missing from the TypeScript SDK.** This PR brings feature parity between the two SDKs.

Traditional evaluation in Weave requires:
- Upfront dataset definition
- Batch processing of all examples
- All scorers defined before evaluation starts

However, many real-world use cases need:
- Streaming data sources (e.g., live production traffic, user sessions)
- Dynamic/conditional scoring (different examples may use different scorers)
- Incremental evaluation results
- Flexible evaluation workflows

**Imperative Evaluation Logger** solves these problems by allowing users to log predictions one at a time, add scores dynamically, and generate summaries on demand.

### Core Components

1. **`EvaluationLogger`** - Main entry point for imperative evaluation
   - Creates root `evaluate` call on construction
   - Manages prediction lifecycle
   - Generates auto-summary or accepts custom summary
   - Tracks all predictions for summary calculation

2. **`ScoreLogger`** - Manages scoring for a single prediction
   - Returned from `logPrediction()`
   - Allows multiple scores to be logged incrementally
   - Creates scorer calls with feedback attachment
   - Finalizes prediction with `finish()`

3. **WeaveObject Classes** (Internal)
   - `Evaluation` - Root evaluation object
   - `Model` - Model wrapper object
   - `Scorer` - Dynamically created per scorer name

4. **Built-in Ops** (Internal)
   - `Evaluation.evaluate` - Root evaluation op
   - `Evaluation.predict_and_score` - Prediction container op
   - `Evaluation.summarize` - Summary generation op
   - `Model.predict` - Prediction op
   - `{scorer_name}.score` - Scorer ops (dynamically created)

### Call Hierarchy
```
evaluate (root)
├── predict_and_score (per prediction)
│   ├── predict (finished immediately)
│   └── {scorer_name}.score (per score, with feedback attachment)
├── predict_and_score (next prediction)
│   └── ...
└── summarize (finalization)
```

## API Usage

```typescript
import {EvaluationLogger} from '@wandb/weave';

// Create evaluation logger
const ev = new EvaluationLogger({
  name: 'my-eval',
  description: 'Evaluating my model',
  dataset: 'my-dataset', // optional
  scorers: ['accuracy', 'f1'], // optional, for warnings only
  model: myModel, // optional
  attributes: {trial_id: 'abc123'} // optional custom attributes
});

// Log predictions incrementally
for (const example of streamingData) {
  const output = await myModel.predict(example);

  // Log prediction and get scorer
  const pred = await ev.logPrediction(example, output);

  // Conditionally add scores
  if (shouldScore(output)) {
    await pred.logScore('accuracy', calculateAccuracy(output));
    await pred.logScore('f1', calculateF1(output));
  }

  // Finalize prediction
  await pred.finish();
}

// Generate summary (auto-calculates mean/true_fraction)
await ev.logSummary();

// Or provide custom summary
await ev.logSummary({
  accuracy: {mean: 0.95},
  custom_metric: 'custom_value'
});
```

#### 1. **Direct Call Management vs. Op System Coordination**

**The Critical Choice:**

The Python SDK achieves imperative evaluation by coordinating the Op system (`__require_explicit_finish`, call stack manipulation, context variables). We **rejected this approach** in favor of **direct call management** using the existing `createCall()`/`finishCall()` API.

**Why Direct Call Management?**
- **40% faster implementation** (6-9 days vs 10-15 days) (Chance: Well the actual time took about 2 days)
- **Zero changes to Op system** - no risk of breaking existing functionality
- **Explicit parent-child relationships** via `createCall(..., parentEntry)`
- **Simple metadata tracking** instead of complex stack manipulation
- **Better debugging** - clear call hierarchy in code, not implicit in framework

**Trade-off:** Requires manual call lifecycle management instead of automatic Op coordination, but this is actually **simpler** and more maintainable.

#### 2. **Metadata-Driven Architecture**

Instead of coordinating Ops, we track call relationships in lightweight metadata:

```typescript
class PredictAndScoreCallMetadata {
  predictAndScoreCall: InternalCall;       // The container call
  predictAndScoreEntry: CallStackEntry;    // Call stack entry
  evaluateEntry: CallStackEntry;           // Parent reference
  predictCallId: string;                   // For feedback attachment
  scores: Record<string, any> = {};        // Accumulated scores
  output: any;                             // Prediction output
  isFinished: boolean = false;             // Lifecycle state
}
```

**Why this matters:**
- **Explicit state tracking** - no hidden framework state
- **Clear parent-child relationships** - set in `createCall()` not via stack manipulation
- **Flexible timing** - finish calls whenever ready, not tied to Op execution
- **Easy debugging** - inspect metadata to see all call relationships

#### 3. **Built-in SDK Ops (Not Dynamic Op Creation)**

We define imperative evaluation Ops as **built-in SDK operations** that are never executed:

```typescript
const evaluationEvaluate = op(
  async function evaluate(evaluation: any, model: any): Promise<any> {
    // Never executed - only for UI display and call structure
    throw new Error('Use EvaluationLogger instead of calling directly');
  },
  {name: 'Evaluation.evaluate'}
);
```

**Why this approach?**
- **Auto-saved by createCall()** - no manual save/publish needed
- **Shared across all instances** - zero per-instance overhead
- **Source code captured** - displayed in Weave UI for documentation
- **Type-safe** - proper TypeScript signatures
- **Internal to module** - not exposed in public API

**Alternative rejected:** Dynamic Op creation per EvaluationLogger instance (expensive, complex lifecycle).

#### 4. **Internal WeaveObject Classes**

Created dedicated `Evaluation`, `Model`, `Scorer` classes **internal to evaluationLogger**:

```typescript
class Evaluation extends WeaveObject {
  dataset?: Dataset<any> | string;
  scorers: string[];
  // Auto-serializes with class_name: 'Evaluation'
}
```

**Why internal classes?**
- **Class name = serialized name** - `class Evaluation` → `class_name: 'Evaluation'` in trace
- **No naming conflicts** - doesn't interfere with existing `Evaluation` class in batch eval
- **Auto-saved by SDK** - `saveWeaveValues()` recursively saves all WeaveObjects in call inputs
- **Flexible attributes** - copy any custom fields (trial_id, etc.) into object

**Key insight:** The SDK's existing `saveWeaveValues()` infrastructure handles all object persistence - we just need to pass WeaveObjects in call inputs.

#### 5. **Scorer Feedback via Finished Predict Call**

Scorers attach results as **feedback** to the already-finished `predict` call:

```typescript
// In logPrediction(): finish predict immediately
await createAndFinishCall(..., modelPredict, ...);
const predictCallId = predictEntry.callId;

// In logScore(): attach feedback to finished predict call
await client.addFeedback(
  predictCallId,        // Target: finished predict call
  scorerEntry.callId,   // Source: scorer call
  scorerRefUri,
  score
);
```

**Why finish predict immediately?**
- **Captures accurate timing** - predict duration reflects actual prediction time
- **Enables feedback attachment** - can only attach feedback to finished calls
- **Matches Python behavior** - scorer results are feedback, not child calls

**Alternative rejected:** Keep predict call open until finish() - would show incorrect durations.

#### 6. **Deferred Finishing for predict_and_score**

The `predict_and_score` call starts in `logPrediction()` but finishes in `finish()`:

```typescript
// In logPrediction():
const startPromise = client.createCall(...);  // Start predict_and_score
await startPromise;
// Return ScoreLogger - call still open

// In finish():
await client.finishCall(..., startPromise);   // Finish with accumulated scores
```

**Why defer finishing?**
- **Accumulate scores** - multiple `logScore()` calls add to `scores` map
- **Correct output** - finish with `{output, scores}` tuple
- **User controls timing** - finish when ready, not tied to framework

**Critical detail:** Must save `startPromise` from `createCall()` and pass to `finishCall()` - required by SDK's call matching logic.

#### 7. **Auto-Summary Generation**

When `logSummary()` called without arguments, auto-generates summary from accumulated scores:

```typescript
private generateAutoSummary(): Record<string, any> {
  // For numeric scores: calculate mean
  if (values.every(v => typeof v === 'number')) {
    summary[scoreName] = {mean: sum / values.length};
  }
  // For boolean scores: calculate true_count and true_fraction
  else if (values.every(v => typeof v === 'boolean')) {
    summary[scoreName] = {
      true_count: trueCount,
      true_fraction: trueCount / values.length
    };
  }
}
```

**Why auto-generate?**
- **Convenience** - users don't need to compute manually
- **Python parity** - matches Python SDK behavior
- **Flexibility** - users can override with custom summary

#### 8. **Async Safety Warning**

Warn if `logSummary()` called with unfinished predictions:

```typescript
const unfinishedCount = this.predictions.filter(p => !p.isFinished).length;
if (unfinishedCount > 0) {
  console.warn(
    `logSummary() called with ${unfinishedCount} unfinished prediction(s). ` +
    `Make sure to call finish() on all ScoreLoggers before calling logSummary().`
  );
}
```

**Why this matters:**
- **Catches user errors** - forgetting to await finish()
- **Prevents incomplete summaries** - scores from unfinished predictions excluded
- **Async timing issues** - JavaScript is single-threaded but async can cause race conditions

### Changes to Existing Code

**`weaveClient.ts`** (sdks/node/src/weaveClient.ts:574-647)
- Added `addFeedback()` method for attaching scorer feedback to predict calls
- Exports `CallStackEntry` type for internal use by evaluationLogger

**`inMemoryTraceServer.ts`** (sdks/node/src/inMemoryTraceServer.ts:173-182)
- Added `feedback` API stub for test infrastructure
- Returns mock feedback IDs for testing

**`index.ts`** (sdks/node/src/index.ts)
- Exports `EvaluationLogger` and `ScoreLogger` classes

**`package.json`**
- Version bump for release

## Testing

Created comprehensive test suite with **10 tests across 5 categories**:

1. **Basic Functionality** (2 tests)
   - Complete evaluation lifecycle
   - Full configuration options with custom attributes

2. **Call Hierarchy** (1 test)
   - Verifies correct parent-child relationships in call tree

3. **Attribute Markers** (2 tests)
   - `IMPERATIVE_EVAL_MARKER` on evaluation calls
   - `IMPERATIVE_SCORE_MARKER` on scorer calls

4. **Summary Generation** (3 tests)
   - Auto-summary calculates `mean` for numeric scores
   - Auto-summary calculates `true_count` and `true_fraction` for boolean scores
   - Custom summary overrides auto-summary

5. **Edge Cases** (3 tests)
   - Prediction with no scores
   - Summary with no predictions
   - Multiple predictions sequentially

**Test Infrastructure:**
- Uses `InMemoryTraceServer` for mocking
- Helper function `getCalls()` with built-in `waitForPendingOperations()`
- All tests pass ✅

## Files Changed

**New Files:**
- `sdks/node/src/evaluationLogger.ts` - 749 lines - Core implementation
- `sdks/node/src/__tests__/evaluationLogger.test.ts` - 303 lines - Test suite

**Modified Files:**
- `sdks/node/src/weaveClient.ts` - Added `addFeedback()` method
- `sdks/node/src/inMemoryTraceServer.ts` - Added feedback API stub
- `sdks/node/src/index.ts` - Export new classes
- `sdks/node/src/weaveObject.ts` - Removed unused code
- `sdks/node/package.json` - Version bump

**Total:** +1,129 lines, -15 lines

### What to Focus On

1. **Async correctness**: Verify the async flow in `logPrediction()` and `logSummary()`
2. **Call hierarchy**: Check that parent-child relationships are correct
3. **Feedback attachment**: Ensure scorer feedback properly attached to predict calls
4. **Auto-summary logic**: Review mean/true_fraction calculations
5. **API ergonomics**: Does the API feel natural for TypeScript users?

### Design Tradeoffs

- **No lock protection on `this.predictions` array**: Safe because JavaScript is single-threaded, but relies on users following correct async patterns
- **Warning-only scorers validation**: Unlike strict enforcement, we only warn about unrecognized scorers (matches Python SDK)
- **No Dataset creation in tests**: Simplified to use string dataset names to avoid testing Dataset functionality

## Related Work

- Python SDK: `weave/evaluation/eval_imperative.py`
- Design doc: [Internal design doc](https://www.notion.so/wandbai/Imperative-Evaluation-Logger-TypeScript-SDK-Design-Document-284e2f5c7ef3804baf18c34b3a7532ee?pvs=12)

## Checklist

- [x] Implementation complete with all core features
- [x] Test suite with 10 comprehensive tests (all passing)
- [x] Async safety warnings added
- [x] TypeScript idioms followed (`this` vs `self`)
- [x] Code cleanup (utility functions, removed redundancies)
- [x] Documentation in code comments
- [x] Exported from main SDK index

### CODE REVIEW

**Comment by chance-wnb** [comment]
* **#5542** <a href="https://app.graphite.dev/github/pr/wandb/weave/5542?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#5543](https://github.com/wandb/weave/pull/5543) <a href="https://app.graphite.dev/github/pr/wandb/weave/5543?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#5555](https://github.com/wandb/weave/pull/5555) <a href="https://app.graphite.dev/github/pr/wandb/weave/5555?utm_source=stack-comment-icon" target="_blank">Graphite</a>) 👈 <a href="https://app.graphite.dev/github/pr/wandb/weave/5542?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: sdks/node/src/weaveObject.ts [hunk 1]
```diff
@@ -50,10 +50,6 @@ export class WeaveObject {
 
   constructor(protected _baseParameters: WeaveObjectParameters) {}
 
-  className() {
```
Comment (chance-wnb) [inline sdks/node/src/weaveObject.ts:53]: No one calls.

File: sdks/node/src/evaluationLogger.ts [hunk 2]
```diff
@@ -0,0 +1,749 @@
...
+      constructor(parameters: ScorerParameters) {
+        super({
+          ...parameters,
+          name: scorerName,
+          scorerName: scorerName,
+        });
+      }
+    }
+
+    // // Override the constructor name to be the scorer name
```
Comment (andrewtruong) [inline sdks/node/src/evaluationLogger.ts]: nit: double comment

File: sdks/node/src/weaveClient.ts [hunk 3]
```diff
@@ -644,9 +644,9 @@ export class WeaveClient {
   ) {
     let inputs: Record<string, any> = {};
 
-    // Add 'self' first if thisArg is a WeaveObject
-    if (thisArg instanceof WeaveObject) {
-      inputs['self'] = thisArg;
+    // Add 'this' first if thisArg is a WeaveObject
+    if (thisArg instanceof WeaveObject && !('this' in inputs)) {
+      inputs['this'] = thisArg;
```
Comment (andrewtruong) [inline sdks/node/src/weaveClient.ts]: Does this work?  Does the UI use the `self` key for anything?

File: sdks/node/src/weaveClient.ts [hunk 4]
```diff
@@ -706,7 +706,8 @@ export class WeaveClient {
     currentCall: CallStackEntry,
     parentCall: CallStackEntry | undefined,
     startTime: Date,
-    displayName?: string
+    displayName?: string,
+    customAttributes?: Record<string, any>
```
Comment (andrewtruong) [inline sdks/node/src/weaveClient.ts]: nit: This should be called `evalAttributes` for consistency

File: sdks/node/src/weaveClient.ts [hunk 5]
```diff
@@ -821,6 +827,54 @@ export class WeaveClient {
...
+    predictCallId: string,
+    scorerCallId: string,
+    runnableRefUri: string,
+    scorerOutput: any
+  ): Promise<string> {
+    // Parse entity and project from projectId (format: "entity/project")
+    const [entity, project] = this.projectId.split('/');
+
+    // Build call URIs in weave:/// format
+    const predictCallUri = `weave:///${entity}/${project}/call/${predictCallId}`;
```
Comment (andrewtruong) [inline sdks/node/src/weaveClient.ts]: maybe this should be factored into a helper

File: sdks/node/src/weaveClient.ts [hunk 6]
```diff
@@ -821,6 +827,54 @@ export class WeaveClient {
...
+  /**
+   * Add feedback (e.g., scorer output) to a call.
+   * Used in imperative evaluation to attach scorer results to predict calls.
+   *
+   * @param predictCallId - ID of the predict call to attach feedback to
+   * @param scorerCallId - ID of the scorer call that generated the feedback
+   * @param runnableRefUri - URI of the scorer (Op or Object ref)
+   * @param scorerOutput - Output of the scorer
+   */
+  public async addFeedback(
```
Comment (andrewtruong) [inline sdks/node/src/weaveClient.ts]: I would rename this to `addScore` as we already have a concept called `Feedback`, and this is specifically for scoring eval predictions

File: sdks/node/src/evaluationLogger.ts [hunk 7]
```diff
@@ -0,0 +1,749 @@
...
+    parameterNames,
+    thisArg,
+    currentEntry,
+    parentEntry,
+    startTime,
+    displayName,
+    attributes
+  );
+
+  await startPromise;
```
Comment (andrewtruong) [inline sdks/node/src/evaluationLogger.ts]: Do you have to block here?

File: sdks/node/src/evaluationLogger.ts [hunk 8]
```diff
@@ -0,0 +1,749 @@
...
+/**
+ * Evaluation WeaveObject for imperative evaluation logging.
+ * Internal to evaluationLogger - not exported from main SDK.
+ *
+ * Named "Evaluation" so that constructor.name serializes to 'Evaluation'.
+ */
+interface EvaluationParameters extends WeaveObjectParameters {
+  dataset?: Dataset<any> | string;
+  scorers?: string[];
+  [key: string]: any; // Allow arbitrary attributes (trial_id, run_id, etc.)
```
Comment (andrewtruong) [inline sdks/node/src/evaluationLogger.ts:109]: Why not explicitly define the params?

File: sdks/node/src/evaluationLogger.ts [hunk 9]
```diff
@@ -0,0 +1,749 @@
...
+// ============================================================================
+
+/**
+ * Evaluation WeaveObject for imperative evaluation logging.
+ * Internal to evaluationLogger - not exported from main SDK.
+ *
+ * Named "Evaluation" so that constructor.name serializes to 'Evaluation'.
+ */
+interface EvaluationParameters extends WeaveObjectParameters {
+  dataset?: Dataset<any> | string;
```
Comment (andrewtruong) [inline sdks/node/src/evaluationLogger.ts:107]: OK for MVP, but you might want to parametrize the types similar to `Evaluation`

File: sdks/node/src/evaluationLogger.ts [hunk 10]
```diff
@@ -0,0 +1,749 @@
...
+  scorers?: string[];
+  [key: string]: any; // Allow arbitrary attributes (trial_id, run_id, etc.)
+}
+
+class Evaluation extends WeaveObject {
```
Comment (andrewtruong) [inline sdks/node/src/evaluationLogger.ts:112]: I think we already have an `Evaluation` inside `evaluation.ts`, why redefine it?

File: sdks/node/src/evaluationLogger.ts [hunk 11]
```diff
@@ -0,0 +1,749 @@
...
+   *
+   * @param scorerName - Name of the scorer (e.g., "accuracy", "f1_score")
+   * @param score - The score value
+   */
+  async logScore(scorerName: string, score: any): Promise<void> {
+    if (this.predMeta.isFinished) {
+      throw new Error('Cannot log score after prediction is finished');
+    }
+
+    const client = getGlobalClient();
```
Comment (andrewtruong) [inline sdks/node/src/evaluationLogger.ts]: we have `requireGlobalClient` for this

File: sdks/node/src/evaluationLogger.ts [hunk 12]
```diff
@@ -0,0 +1,749 @@
...
+      scorerRefUri,
+      score
+    );
+
+    // Store score in metadata
+    this.predMeta.scores[scorerName] = score;
+  }
+
+  /**
+   * Finish this prediction.
```
Comment (andrewtruong) [inline sdks/node/src/evaluationLogger.ts]: You mean scorer?

File: sdks/node/src/evaluationLogger.ts [hunk 13]
```diff
@@ -0,0 +1,749 @@
...
+   * Finalizes the predict_and_score call with accumulated scores.
+   * Note: The predict call is finished by EvaluationLogger, not here.
+   */
+  async finish(): Promise<void> {
+    if (this.predMeta.isFinished) {
+      return; // Already finished
+    }
+
+    const client = getGlobalClient();
```
Comment (andrewtruong) [inline sdks/node/src/evaluationLogger.ts]: use `requireGlobalClient`

File: sdks/node/src/evaluationLogger.ts [hunk 14]
```diff
@@ -0,0 +1,749 @@
...
+      undefined,
+      evaluateEntry,
+      undefined, // No parent (root call)
+      startTime,
+      undefined, // displayName
+      {...IMPERATIVE_EVAL_MARKER, ...this.evalAttributes} // Merge eval attributes
+    );
+    await evaluateStartPromise;
+
+    this.evaluateCall = evaluateCall;
```
Comment (andrewtruong) [inline sdks/node/src/evaluationLogger.ts:540]: What is the difference between this and `evaluateStartPromise`?

File: sdks/node/src/__tests__/evaluationLogger.test.ts [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+    const evalLogger = new EvaluationLogger({name: 'test-eval'});
+
+    const scoreLogger = await evalLogger.logPrediction(
+      {input: 'test'},
+      'output'
+    );
+
+    await scoreLogger.logScore('accuracy', 0.95);
+    await scoreLogger.logScore('f1', 0.88);
+    await scoreLogger.finish();
```
Comment (andrewtruong) [inline sdks/node/src/__tests__/evaluationLogger.test.ts]: Should the user need to await all of these?  I think it makes more sense that you just need to wait for the group, not individually

File: sdks/node/src/weaveClient.ts (hunk 3 above)
Comment (chance-wnb) [reply sdks/node/src/weaveClient.ts]: I haven't seen anything broken UI-wise yet.

File: sdks/node/src/weaveClient.ts (hunk 4 above)
Comment (chance-wnb) [reply sdks/node/src/weaveClient.ts]: Hmm, I see what you mean. 

`createCall()` theoretically is capable of serving consumers other than the eval scenarios. So I think having a natural name here is fine. Since this is nit, I will keep it as is.

File: sdks/node/src/evaluationLogger.ts (hunk 8 above)
Comment (chance-wnb) [reply sdks/node/src/evaluationLogger.ts:109]: User has the freedom to specify any parameter as they desire. The parameters will be displayed as metadata in the UI. User will eventually decide what metadata they want to see. So I guess we don't need to dictate anything here.

File: sdks/node/src/evaluationLogger.ts (hunk 9 above)
Comment (chance-wnb) [reply sdks/node/src/evaluationLogger.ts:107]: Maybe I don't follow, but I think the parameters are flexible parameters. Can you show a code example on what you expect it to be?

File: sdks/node/src/weaveClient.ts (hunk 4 above)
Comment (andrewtruong) [reply sdks/node/src/weaveClient.ts]: Oh sorry, I didn't see that this was the generic `createCall` -- in that case I think this should just be called `attributes`

File: sdks/node/src/evaluationLogger.ts (hunk 10 above)
Comment (chance-wnb) [reply sdks/node/src/evaluationLogger.ts:112]: Great question. 

The new classes are for serialization purpose only, I don't want to pull in any other behaviors of a class. According to the logic of [`saveObject()`](https://github.com/wandb/weave/blob/e75e8fcff4f211330afce3a7ecdf6781d512d91b/sdks/node/src/weaveClient.ts#L422) it only looks at the class name and its enumerable properties. For cleanness, I want to have a simple class that is away from any potential intricacies that might come from the classic Evaluation class. 

These object are automatically serialized when they appear in the input of a call. Having this natural serialization process is ideal, that's why dummy classes are created.

File: sdks/node/src/evaluationLogger.ts (hunk 14 above)
Comment (chance-wnb) [reply sdks/node/src/evaluationLogger.ts:540]: The evaluateCall (InternalCall) is a wrapper of CallSchema, which aims to add some behaviors. 
`evaluateStartPromise` is just a signal to gauge the timing on `createCall()`.

Since `evaluateStartPromise` does not live inside evaluateCall, so I need to keep track of them separately.

File: sdks/node/src/__tests__/evaluationLogger.test.ts (hunk 15 above)
Comment (chance-wnb) [reply sdks/node/src/__tests__/evaluationLogger.test.ts]: That's a great insight! No they shouldn't need to wait on these. We should strive for fire-n-forget. 

It does make things a bit trickier: since we need to ensure the scorings are all done before finishing the `predict_and_call`. Added a promise capturing for that. 

This part has been redone, please kindly help review again! Thanks!

File: sdks/node/src/evaluationLogger.ts (hunk 12 above)
Comment (chance-wnb) [reply sdks/node/src/evaluationLogger.ts]: yeah, reworded a bit for clarity.

File: sdks/node/src/evaluationLogger.ts (hunk 7 above)
Comment (chance-wnb) [reply sdks/node/src/evaluationLogger.ts]: Nope, removed. Thanks!

File: sdks/node/src/weaveClient.ts (hunk 6 above)
Comment (chance-wnb) [reply sdks/node/src/weaveClient.ts]: Indeed. renamed as such.

File: sdks/node/src/weaveClient.ts (hunk 5 above)
Comment (chance-wnb) [reply sdks/node/src/weaveClient.ts]: I added a CallRef class for that, please let me know if the shape is good.

File: sdks/node/src/weaveClient.ts (hunk 4 above)
Comment (chance-wnb) [reply sdks/node/src/weaveClient.ts]: Yep, renamed to `attributes` for now.

File: sdks/node/src/weaveClient.ts (hunk 3 above)
Comment (chance-wnb) [reply sdks/node/src/weaveClient.ts]: Sorry Andrew, somehow after the new iteration I start to see something broken. I have to have this PR to fix (core side): https://github.com/wandb/core/pull/35201

Do you feel this is risky? I so, I am happy to revert back to using `self`.

Comment (andrewtruong) [reply sdks/node/src/weaveClient.ts]: I think the client would break on older servers.  The safest option would be to use `self`, then switch back to `this` next month (to align with the next server release cycle`)

Comment (chance-wnb) [reply sdks/node/src/weaveClient.ts]: Indeed, I have reverted back to use self.

File: sdks/node/src/evaluationLogger.ts (hunk 9 above)
Comment (andrewtruong) [reply sdks/node/src/evaluationLogger.ts:107]: For example, the regular `Evaluation` is parametrized with `Evaluation<R, E, M>`

Since this is just metadata I don't think this is a huge issue.

Comment (chance-wnb) [reply sdks/node/src/evaluationLogger.ts:107]: Ah, I see what you mean. I envision these part to be written after integrating your Eval V2 API. I will address these then.

File: sdks/node/src/__tests__/evaluationLogger.test.ts [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+    scoreLogger1.logScore('passed', true);
+    await scoreLogger1.finish();
+
+    const scoreLogger2 = await evalLogger.logPrediction({input: '2'}, 'out2');
+    scoreLogger2.logScore('passed', false);
+    await scoreLogger2.finish();
+
+    const scoreLogger3 = await evalLogger.logPrediction({input: '3'}, 'out3');
+    scoreLogger3.logScore('passed', true);
+    await scoreLogger3.finish();
```
Comment (andrewtruong) [inline sdks/node/src/__tests__/evaluationLogger.test.ts:205]: I don't think the user should need to call await here, except maybe at the very end.

File: sdks/node/src/__tests__/evaluationLogger.test.ts [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+    const evalLogger = new EvaluationLogger({name: 'test-eval'});
+
+    for (let i = 0; i < 3; i++) {
+      const scoreLogger = await evalLogger.logPrediction(
+        {input: `test${i}`},
+        `output${i}`
+      );
+      scoreLogger.logScore('accuracy', 0.9 + i * 0.05);
+      await scoreLogger.finish();
+    }
```
Comment (andrewtruong) [inline sdks/node/src/__tests__/evaluationLogger.test.ts:181]: Why do you need to block here?

File: sdks/node/src/__tests__/evaluationLogger.test.ts [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+  beforeEach(() => {
+    traceServer = new InMemoryTraceServer();
+    initWithCustomTraceServer(projectId, traceServer);
+  });
+
+  // Test: Basic flow - logPrediction → logScore → finish → logSummary
+  test('complete evaluation lifecycle', async () => {
+    const evalLogger = new EvaluationLogger({name: 'test-eval'});
+
+    const scoreLogger = await evalLogger.logPrediction(
```
Comment (andrewtruong) [inline sdks/node/src/__tests__/evaluationLogger.test.ts:35]: do you need to await here?

**Comment by andrewtruong** [comment]
I think this is good for an MVP.  There are a few places where I think we don't need to be awaiting anything, but we can fix that in a future iteration.
